### ISSUE TITLE
Probable concurrency bug in GelfChunkAggregator.checkForCompletion ()

### ISSUE THREAD

**Issue by uttg** [Graylog2/graylog2-server#1561]
Hello,

I investigated on #1544 a little bit further.

For a message with 6 chunks I have the following behaviour:
* *reception of the first chunks...*
* reception of chunk number 3
  * checkForCompletion detects that we have every chunks
  * it fails to construct a full message. Chunks 4, 5 and 6 are not yet in entry.payloadArray
* reception of chunk number 6
  * checkForCompletion detects that we have every chunks
  * it fails to construct a full message. Chunks 1, 2 and 3 are not yet in entry.payloadArray

I noticed that the test [if (chunkWatermark == sequenceCount)](https://github.com/Graylog2/graylog2-server/blob/99d3e248090547114cb8d5b0ad4f07cf9abbde27/graylog2-inputs/src/main/java/org/graylog2/inputs/codecs/GelfChunkAggregator.java#L144) may return true multiple times for the same message.

Somehow, [incrementAndGet()](https://github.com/Graylog2/graylog2-server/blob/99d3e248090547114cb8d5b0ad4f07cf9abbde27/graylog2-inputs/src/main/java/org/graylog2/inputs/codecs/GelfChunkAggregator.java#L141) returns the same value multiple times event though the operation should be atomic.

In graylog.conf we have:
```
processor_wait_strategy = blocking
inputbuffer_processors = 2
inputbuffer_wait_strategy = blocking
```

**Comment by tomtomgod** [comment]
This is probably not the cause, but the method equals() of ChunkEntry will always return false.
Line 255 tries to compare two AtomicInteger:
```c#
 if (!chunkSlotsWritten.equals(that.chunkSlotsWritten)) return false;
```
This comparison is always false. But I don't see how that would affect the rest of the code.

**Comment by tomtomgod** [comment]
I found the problem: it's in the GELF Client we are using :confounded:
Gelf4Net's UDP appender generates consecutive messages with the same messageIds when logging at high frequency.

I don't know what you want to do with my comments on GelfChunkAggregator. Maybe add some defensive code when receiving the same chunk number for the same messageId. Something like:

```c#
if (!entry.payloadArray.compareAndSet(chunk.getSequenceNumber(), null, chunk)) {
    log.error("Duplicated messageId {}", chunk.getId());
    return null;
}
```

**Comment by kroepke** [comment]
I agree with https://github.com/Graylog2/graylog2-server/issues/1561#issuecomment-156519636 that we should add some defensive code for that scenario because it is either a programming error in the client, or a maliciously crafted packet.

I'm not overly sure about the error log level, because that can easily flood the server log, too.
At the very least we should add an additional counter metric, I think.

What do you think @joschi ?

**Comment by joschi** [comment]
> I'm not overly sure about the error log level, because that can easily flood the server log, too.

This error *should not* happen often and if it does it's a sign of a broken client, so I think it's ok to flood the logs with those messages if this error occurs.

> At the very least we should add an additional counter metric, I think.

Yes.

**Comment by kroepke** [comment]
Agreed on the error level.

**Comment by kroepke** [comment]
fixed by referenced PR, which will be in 1.3 and 2.0.

Many thanks for the investigation!
